### PR TITLE
Add support for the `variant` operator in the type and effect system

### PR DESCRIPTION
**Pull request by shonfeder**
Addresses the first half of #244.

The `variant` operator is the builtin operator introduced in #1206 that injects a value into a sum type. I.e., `type T = A(int) | B; val a : T = variant("A", 42)`.

This PR adds support in the type system for checking and inferring the type of this operator. 

The most significant change is the introduction of a new constraint `isDefined(t)` which holds when the type schema `t` unifies with a defined type. This is required and foretold by the RFC001, approved in #1062. It ensures that users cannot introduces open-rowed variants, since we have decided that the complexity of allowing open, anonymous variant types is undesirable.

This PR also corrects the logic around generating the sum type injector operators introduced in #1206.

Please see the commit messages for context. Review by commit my also be helpful. 

- [x] Tests added for any new code
- [x] Documentation added for any new functionality
- [x] Entries added to the respective `CHANGELOG.md` for any new functionality
- [x] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

File: quint/test/types/constraintSolver.test.ts [hunk 1]
```diff
@@ -192,6 +234,33 @@ describe('unify', () => {
     result.map(subs => assert.deepEqual(substitutionsToString(subs), '[ a |-> int, b |-> bool ]'))
   })
 
+  it('unifies sum-type', () => {
+    // We do not provide
```
Comment (bugarela) [inline quint/test/types/constraintSolver.test.ts:238]: Seems like there is something missing in this comment

File: quint/test/types/constraintSolver.test.ts [hunk 2]
```diff
@@ -192,6 +234,33 @@ describe('unify', () => {
...
+    assert.isTrue(result.isRight())
+    result.map(subs => assert.deepEqual(substitutionsToString(subs), '[ a |-> str, b |-> int ]'))
+  })
+
+  // it('unifies sum type with open row', () => {
+  //   const result = unify(table, parseTypeOrThrow('A(a)'), parseTypeOrThrow('A(int) | B(str)'))
+
+  //   assert.isTrue(result.isRight())
+  //   result.map(subs => assert.deepEqual(substitutionsToString(subs), '[]'))
+  // })
```
Comment (bugarela) [inline quint/test/types/constraintSolver.test.ts:262]: Should this be uncommented or removed?

File: quint/src/types/specialConstraints.ts [hunk 3]
```diff
@@ -179,3 +180,49 @@ export function itemConstraints(
...
+  if (labelExpr.kind !== 'str') {
+    return left(
+      buildErrorLeaf(
+        `Generating variant constraints for ${args.map(a => expressionToString(a[0]))}`,
+        `Variant label must be a string expression but is ${labelType.kind}: ${expressionToString(labelExpr)}`
+      )
+    )
+  }
+  // We now know the label must be a quint string
+  const labelStr = labelExpr as QuintStr
```
[the remark this answers is not in the document]
Comment (shonfeder) [reply quint/src/types/specialConstraints.ts:210]: You are right! That cleans things up a bit. Thanks :)

File: quint/test/types/constraintSolver.test.ts (hunk 1 above)
Comment (shonfeder) [reply quint/test/types/constraintSolver.test.ts:238]: Yep. But I don't know what. I've just removed it.

File: quint/test/types/constraintSolver.test.ts (hunk 2 above)
Comment (shonfeder) [reply quint/test/types/constraintSolver.test.ts:262]: Removed! It was part of debugging. Thanks :)

**Comment by shonfeder** [comment]
Thanks for the review and for catching my oversights.

**Pull request merged**